### ISSUE TITLE
Missing method that is public but defined on a non-public superclass

### ISSUE THREAD

**Issue by fidothe** [jruby/jruby#6197]
This is the same issue seen in #5885, but now extends to OpenJDK 8 as well as OpenJDK 12 (these are all I've tested on) under > 9.2.10.

**Environment Information**
JRuby 9.2.10.0, 9.2.11.1 invoked with --dev --debug

x86_64, macOS 10.15.4 (Darwin Kernel Version 19.4.0)

**Expected**
Calling the `setInitialMode` on `net.sf.saxon.s9api.Xslt30Transformer` (which is defined on the abstract class `net.sf.saxon.s9api.AbstractXsltTransformer`) should not explode with a `NoMethodError`

**Actual**
On 9.2.8, 9.2.10, and 9.2.11 under OpenJDK 12 this goes boom. On 9.2.8 under OpenJDK 8, it doesn't explode and works as expected (as reported in #5585). On 9.2.10 and 9.2.11 under OpenJDK 8 it fails in the same way as it does under OpenJDK 12.

I'll add a test case and see if I can reproduce under master too.

**Comment by headius** [comment]
We did not get this fixed for 9.2.12 but there's a workaround use `java_alias` to explicitly bind the missing methods:

```ruby
class net::sf::saxon::s9api::AbstractXsltTransformer
  java_alias :setInitialMode, :setInitialMode, [<parameter types for setInitialMode here>]
end
```

It's not pretty but it works for now.

**Comment by headius** [comment]
The general issue here, to help searches and users that run into this in the future, is that we are completely ignoring the methods on non-public classes even if they would be visible from a child class. This is the simple form that represents the original reported issue:

```java
abstract Foo {
  public void foo() {}
}

public Bar extends Foo {}
```

The `foo` method can't be called against a `Foo` instance since `Foo is not public, but it can be called against a `Bar` instance, since both the method and the target class are public in that case. Because `Foo` is not public, we do not bind *any* methods from it, so it ends up being missing from `Bar` instances.

The attached PR attempts to fix this, but it has some problems.

**Comment by Morozzzko** [comment]
I've met this issue while trying to integrate github.com/structurizr/java and JRuby.

Top-level issue: a method was missing even though it existed in a .jar and was accessible via Java and Kotlin. It returned a simple `NoMethodError`

When I tried to introspect the class via `#methods` and `#instance_methods` I couldn't see _any_ methods which were defined in a class.

I've asked @ [matrix](https://matrix.to/#/!vyEDBdbmzqApWaugdd:matrix.org/$1594159937527482mpvQs:matrix.org?via=matrix.org&via=sibnsk.net&via=hackerspaces.be), and it turns out that this is the same problem. I've had a non-public class in the ancestry chain, and none of the methods from it were exposed

I've built a [gist](https://gist.github.com/Morozzzko/f0b7796e69eadbb9997dc57a2b3d4e05) to reproduce it. Here's a minimized (yet working) content:

test.sh:
```sh
#!/usr/bin/env sh

mvn dependency:get -Dartifact=commons-logging:commons-logging:1.2 -Ddest=.
mvn dependency:get -Dartifact=com.structurizr:structurizr-core:1.4.7 -Ddest=.

jruby test.rb

echo "Inspecting methods from Java::ComStructurizrModel::ModelItem's jar"

javap -classpath structurizr-core-1.4.7.jar -p "com.structurizr.model.ModelItem"
```

test.rb: 

```ruby

require 'java'

require_relative 'commons-logging-1.2.jar'
require_relative 'structurizr-core-1.4.7.jar'

workspace = ::Java::ComStructurizr::Workspace.new('Workspace', 'TestWorkspace')

model = workspace.get_model

system = model.add_software_system('Some Software System', 'Description')
system.addTags('hello') # => NoMethodError
```

When we run the script, we can observe that `javap` outputs a couple of methods, yet `::Java::ComStructurizr::Workspace.instance_methods` doesn't show _any_ of them. 

I've tried the solution above and it works fine.

**Comment by headius** [comment]
The specific solution for @Mogztter's case is below:

```ruby
class com::structurizr::model::ModelItem
  java_alias :addTags, :addTags, [java.lang.String[]]
end
```

**Comment by headius** [comment]
This slipped under the radar and was not fixed in 9.2.12 (and it's possible it would have been too invasive anyway) so I've marked it for 9.3. I consider it a pretty large issue that we aren't binding methods in Ruby that would be callable from Java.

**Comment by headius** [comment]
Also, thank you @Mogztter for the neat reproduction... it's a great real-world case that shows why we need to get this fixed!

**Comment by fidothe** [comment]
> We did not get this fixed for 9.2.12 but there's a workaround use `java_alias` to explicitly bind the missing methods:

Aha, that's great. Thank you very much for that.

> It's not pretty but it works for now.

It does indeed, and means I can cut a release that works with any 9-series JRuby again!

**Comment by headius** [comment]
@fidothe Probably not going to see a fix in 9.2.13 either... this is a tricky one to fix because our class structure doesn't quite work as-is with this kind of lookup. We'll keep evaluating some options and get it fixed for 9.3.

**Comment by fidothe** [comment]
The workaround is fine for my use case, no stress on my end now (and thanks again for suggesting it, I probably wouldn't have arrived at it independently):

All the Java method calls people make via my library are hidden underneath my wrapper layer, so I can control that and if I need to add a new method to that workaround, I understand how to do it, and the tests will stop me missing things...
